Support "system" theme setting (light/dark theme depending on user system preference)

renchap opened the pull request:
This is built on the work by @nshki in #20107

I rebased it, made the emoji selection code simpler and ensure it uses a consistent theme:
- in the web UI, the user setting is used, falling back to the instance setting, falling back to the `settings.yml` default
- in the embedded view, the instance setting is used, falling back to the `settings.yml` default. It ignores the user setting, as we want it to look the same when previewing it and when it is embedded in an external website
- the static error page always uses the `settings.yml` theme, because it cant access the DB/cache when generated, but still allows an admin to change it if they really want to

I tested:
- that you can still select a theme and it works (including special emoji handling)
- that when you configure the system theme as a user, everything works as expected (including emoji)
- that when an admin change the server-wide setting, the embedded page uses this theme
- the `config/settings.yml` theme is used for the static error HTML page
- that changing the theme in `config/settings.yml` still works if there is no server-wide setting set

This does not change the theme for any user who configured one, or on any server that overrode it in Admin => Settings => Appearance => Theme

It will switch users to the new system theme if they never saved their preferences AND their admin never changed the settings.

Closes #20107
Closes #10976
Fixes MAS-82

nshki committed:
Show new "System settings" option in site themes

Use light/dark styles by default or when selected

Have 'system' be default theme

This is so that not-logged-in users will have their system preferences
respected out of the box.

Have emoji border logic/tests support light/dark

Update English copy for system theme

Use "Automatic (use system theme)" per suggested commented in GitHub.

Fix `precompile:assets` error

This bundles the new application stylesheet pack tags into a shared
partial that's then used across multiple layouts. This wasn't necessary
before because only one pack tag was being rendered based on the user's
current theme. Now we must conditionally render different pack tags with
media queries if applicable.

Fix `Warden::Proxy` error during build

Follow-up from last commit. This attempts to fix an issue where
`Warden::Proxy` isn't available during the precompilation step. Instead
of calling out to `current_theme` in layouts where it didn't previously,
this reverts the rendering of stylesheet pack tags to explicit
stylesheets in `layouts/embedded.html.haml` and
`layouts/error.html.haml`.

Fix up a bad merge

The `replacement` variable declaration was completely missing. This
commit adds it back in.

Resolve linting warning

Using single quotes instead of double quotes where interpolation isn't needed.

renchap committed:
Simplify emoji selection code

renchap committed:
Add the `system` theme in the available themes everywhere

Otherwise it will not be recognized as a valid theme when trying to determine the style tags to inject

renchap committed:
Respect theme server setting for embedded view and error page

renchap committed:
Fix JS tests

renchap commented:
So the production build is failing because I am using `current_theme` in the error page header, which calls `Setting`, which tries to read the setting from Redis (or the DB).

We could either only use the `config/settings.yml` for this file (which would cause the error page to not use the admin configured theme if any, which is what currently happens), or require redis (and DB?) when pre-compiling assets.

mjankowski commented:
I think for the purposes of this PR we should do the smallest thing that will work, and preserve the existing behavior of just using the static file configuration and not pull redis/DB into the asset compile process. I think that would mean that the change here is that the new system-setting-honoring theme would be used for those?

If we want to support the admin settings overriding those defaults for the static pages, do that as a separate change either after or before this theme behavior change.

mjankowski commented on config/locales/en.yml:
```diff
@@ -1769,6 +1769,7 @@ en:
     contrast: Mastodon (High contrast)
     default: Mastodon (Dark)
```
Probably not in this PR - but a possible future improvement here would be a rename in this area so that the collection becomes: `[high-contrast, mastodon-light, mastodon-dark, system]` (or something along those lines).

I could definitely see having a theme named "default" - which is not the default! - becoming confusing.

We'd need a migration to update settings, and some code renames as well.

renchap replied:
We could do this, but it will need to be in another PR with a migration in the various places where this is used (instance settings and user settings). I suspect it might break other things that rely on the `default` string somewhere (like the emoji code at the moment is relying on `mastodon-light`)

mjankowski replied:
Yes, definitely out of scope for this PR, and noted that we'd have to be diligent on rename.

renchap committed:
Do not require Redis/DB when precompiling assets, and use a helper rather than a view

A helper is easier to test here. Also `theme` is now a parameter, so:
- in the web UI, the user setting is used, falling back to the instance setting, falling back to the `settings.yml` default
- in the embedded view, the instance setting is used, falling back to the `settings.yml` default. It ignores the user setting, as we want it to look the same when previewing it and when it is embedded in an external website
- the static error page always uses the `settings.yml` theme, because it cant access the DB/cache when generated, but still allows an admin to change it if they really want to

renchap commented:
Pushed a new commit, with a helper as it is easier to test.

I changed the theme selection so:
- in the web UI, the user setting is used, falling back to the instance setting, falling back to the `settings.yml` default
- in the embedded view, the instance setting is used, falling back to the `settings.yml` default. It ignores the user setting, as we want it to look the same when previewing it and when it is embedded in an external website
- the static error page always uses the `settings.yml` theme, because it cant access the DB/cache when generated, but still allows an admin to change it if they really want to

vmstan commented:
I have it running on vmst.io now and working well, from what I've seen so far.

mjankowski commented:
> Pushed a new commit, with a helper as it is easier to test.

I think this is good now -- in my original suggestion I'd imagine all the layouts calling the same helper and was imaging a lot more logic in there. What you've done here is probably simpler/better -- each layout is more explicit about what it is doing, and the amount of logic in helper method is reduced. Still may make sense to put some lightweight spec there for regressions, but I think changes are good.

ClearlyClaire reviewed (dismissed):
Looks good to me overall, with a little nitpick about the naming of one parameter (see inline comment)

renchap committed:
Rename function argument

renchap commented on app/javascript/mastodon/features/emoji/emoji.js:
```diff
@@ -17,8 +17,13 @@ const emojiFilenames = (emojis) => {
+ * @param {"light" | "dark" } theme
+ * @returns {string}
+ */
+const emojiFilename = (filename, theme) => {
```
Changed to `colorScheme`

Fffjjy reviewed (commented):
Fix and replace
